airgen/src/lib.rs

```diff
@@ -295,11 +309,128 @@ impl<'a> ASMPILConverter<'a> {
                         id: d.operation.id.id.clone(),
                         params: d.operation.params.clone(),
                     },
-                    selector_idx,
+                    // this will be set later, after compatible links are merged
+                    selector_idx: None,
                 })
                 .unwrap()
                 .clone(),
             is_permutation,
         }
     }
+
+    /// Process each link and then combine compatible links.
+    /// Links can be merged iff:
+    /// - they originate from the same machine instance
+    /// - they target the same instance.operation
+    /// - they are of the same kind (permutation/lookup)
+    /// - their flags are mutually exclusive
+    /// Right now we only consider links from different instructions,
+    /// as a single instruction can be active at a time
+    fn process_and_merge_links(&self, defs: &[LinkDefinition]) -> Vec<Link> {
+        /// Helper struct to group links that can potentially be merged.
+        /// Besides these being equal, the links must be mutually exclusive (e.g., come from different instructions)
+        #[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Debug)]
+        struct LinkInfo {
+            from: Location,
+            to: Location,
+            operation: Operation,
+            is_permutation: bool,
+        }
+
+        // resulting links
+        let mut links = vec![];
+
+        // idea here is:
+        // - group by LinkInfo
+        // - inside each group, separate links into sets of with mutually exclusive flags (that is, from different instructions)
+        let mut grouped_links: BTreeMap<LinkInfo, Vec<BTreeMap<Expression, Link>>> =
+            Default::default();
+        for link_def in defs.iter() {
```

Comment:
I have the impression that this would be better readable if you
- partition the links into groupable and non-groupable, this could be as simple as mapping all links into pairs of `(Option<LinkIndo>, Link)`, where non-groupable are mapped to None in the first component.
- then you collect that into a BTreeMap as you already do and then
- process the None-group by just returning it and
- process the other groups by merging